File: Haskell/src/Cat.hs
module Cat where

import           Printable


-- domain classes
--
data Cat = Cat
  { name  :: String
  , age   :: Int
  , color :: String
  } deriving (Eq, Show)


-- type class instances for doamin classes
--
instance Printable Cat where
    format cat = "Cat {name=" ++ name cat ++ ", age=" ++ show (age cat) ++ ", color=" ++ color cat ++ "}"

File: Scala/src/main/scala/typeclasses/_01b_my_printable_noimports/domain/Cat.scala
package typeclasses._01b_my_printable_without_imports.domain

import typeclasses._01b_my_printable_without_imports.libPrintable._

final case class Cat(name: String, age: Int, color: String)

object Cat {

  implicit val catPrintable: Printable[Cat] = new Printable[Cat] {
    override def format(cat: Cat): String = {
      val name  = cat.name.format
      val age   = cat.age.format
      val color = cat.color.format
      s"$name is a $age year-old $color cat."
    }
  }
}
